src/OrasProject.Oras/Content/ReadOnlyGraphStorageExtensions.cs

```diff
@@ -0,0 +1,167 @@
+        }
+
+        await Task.WhenAll(copyTasks).ConfigureAwait(false);
+    }
+
+    /// <summary>
+    /// Represents a node in the graph traversal with its depth information.
+    /// Used during the depth-first search to find root nodes.
+    /// </summary>
+    internal record NodeInfo(Descriptor Node, int Depth);
+
+    /// <summary>
+    /// Finds the root nodes reachable from the given node through a depth-first search.
+    /// </summary>
+    /// <param name="src">The source graph storage.</param>
+    /// <param name="node">The descriptor identifying the starting node.</param>
+    /// <param name="opts">Options for the extended copy operation.</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>A list of descriptors representing the root nodes.</returns>
+    internal static async Task<List<Descriptor>> FindRootsAsync(
+        this IReadOnlyGraphStorage src,
+        Descriptor node,
+        ExtendedCopyGraphOptions opts,
+        CancellationToken cancellationToken = default)
+    {
+        var visited = new HashSet<BasicDescriptor>();
+        var rootMap = new Dictionary<BasicDescriptor, Descriptor>();
+
+        void AddRoot(BasicDescriptor key, Descriptor val)
+        {
+            if (!rootMap.ContainsKey(key))
+            {
+                rootMap[key] = val;
+            }
+        }
+
+        opts.FindPredecessors ??= async (src, descriptor, cancellationToken) =>
+        {
+            return await src.GetPredecessorsAsync(descriptor, cancellationToken).ConfigureAwait(false);
+        };
+
+        var stack = new Stack<NodeInfo>();
+
+        // push the initial node to the stack, set the depth to 0
+        stack.Push(new NodeInfo(node, 0));
+        while (stack.TryPop(out var current))
+        {
+            var currentNode = current.Node;
+            var currentKey = currentNode.BasicDescriptor;
+            if (visited.Contains(currentKey))
+            {
+                continue;
+            }
+            visited.Add(currentKey);
+
+            if (opts.Depth > 0 && current.Depth == opts.Depth)
+            {
+                AddRoot(currentKey, currentNode);
+                continue;
+            }
+
+            IEnumerable<Descriptor> predecessors;
+            try
+            {
+                predecessors = await opts.FindPredecessors(src, currentNode, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidOperationException("FindPredecessors operation failed", ex);
+            }
+
+            var predecessorList = predecessors.ToList();
+
+            if (predecessorList.Count == 0)
+            {
+                AddRoot(currentKey, currentNode);
+                continue;
+            }
+
+            foreach (var predecessor in predecessorList)
+            {
```

Comment:
Consider reversing the loop order on `predecessorList` to match the standard recursive DFS traversal order (left-to-right pre-order).

**Rationale**
The current forward foreach followed by `stack.Push` processes the last predecessor first due to LIFO stack behavior, yielding reversed visit order compared to recursive DFS (e.g., A → C → B instead of A → B → C).

To match recursion exactly, iterate in reverse so the stack pops predecessors left-to-right.

```suggestion
            // Push predecessors in reverse order to match recursive DFS (left-to-right visit order)
            for (int i = predecessorList.Count - 1; i >= 0; i--)
            {
                var predecessor = predecessorList[i];
```

This ensures identical traversal sequence to the recursive version.